File: src/state.rs
use crate::prelude::*;

pub(crate) struct State {
    pub(crate) ic: InitialConditions,
    pub(crate) n: usize,
    pub(crate) h: f64,
    pub(crate) dt: f64,
    pub(crate) alpha: f64,
    pub(crate) gamma: f64,
    pub(crate) rho: Vector,
    pub(crate) p: Vector,
    pub(crate) u: Vector,
    pub(crate) e: Vector,
    pub(crate) E: Vector,
    pub(crate) c: Vector,
}

impl State{
    pub(crate) fn new(n: usize, h: f64, alpha: f64, dt:f64) -> Self {
        
		let gamma = 1.4;

		let ic = InitialConditions::new(gamma);

		let mut rho = Vector::zeros(n);
		let mut p = Vector::zeros(n);
		let mut u = Vector::zeros(n);
		let mut e = Vector::zeros(n);
		let mut E = Vector::zeros(n); // e + 1/2 u^2
		let mut c = Vector::zeros(n);

		// the discontinuity should be exactly in the middle
		let half_n = (n-1) / 2;

        rho.slice_mut(s!(0..half_n)).fill(ic.rhoL);
        rho.slice_mut(s!(half_n..n)).fill(ic.rhoR);

        p.slice_mut(s!(0..half_n)).fill(ic.pL);
        p.slice_mut(s!(half_n..n)).fill(ic.pR);

        u.slice_mut(s!(0..half_n)).fill(ic.uL);
        u.slice_mut(s!(half_n..n)).fill(ic.uR);

        e.slice_mut(s!(0..half_n)).fill(ic.eL);
        e.slice_mut(s!(half_n..n)).fill(ic.eR);

        c.slice_mut(s!(0..half_n)).fill(ic.cL);
        c.slice_mut(s!(half_n..n)).fill(ic.cR);

        for i in 0..half_n {
            E[i] = ic.eL + 0.5 * ic.uL.powf(2.);
        }
        for i in half_n..n{
            E[i] = ic.eR + 0.5 * ic.uR.powf(2.);
        }

        Self { rho, p, u, e, E, c, alpha, h, n, ic, dt, gamma }
    }

    pub(crate) fn check_state(&self, location: &str) {
        if self.rho.has_negs() {
            panic!("rho has negative values at {location}")
        }
        if self.rho.has_negs() {
            panic!("rho has nans at {location}")
        }
        if self.rho.has_zero() {
            panic!("rho has zero at {location}")
        }

        if self.e.has_nans() {
            dbg!(&self.e);
            panic!("entropy has nans at {location}")
        }
        if self.e.has_negs() {
            dbg!(&self.e);
            panic!("entropy has negatives at {location}")
        }
    }
}

pub(crate) struct InitialConditions{
    pub(crate) uR: f64,
    pub(crate) uL: f64,
    //
    pub(crate) pR: f64,
    pub(crate) pL: f64,
    //
    pub(crate) rhoR: f64,
    pub(crate) rhoL: f64,
    //
    pub(crate) eR: f64,
    pub(crate) eL: f64,
    //
    pub(crate) cR: f64,
    pub(crate) cL: f64,
    //
}

impl InitialConditions {
    fn new(gamma: f64) -> Self {
		let uL = 0.;
		let uR = 0.;
		let pL = 10f64.powf(5.);
		let pR = 10f64.powf(3.);
		let rhoL  = 1.0;
		let rhoR = 0.01;
		// from the paper p = (gamma - 1) rho e
		// so you can calculate these
		let eL = pL / ((gamma - 1.) * rhoL);
		let eR = pR / ((gamma - 1.) * rhoR);

		// given in the paper again
		let cL = (gamma * pL/ rhoL).sqrt();
		let cR = (gamma * pR / rhoR).sqrt();

        Self {uR, uL, pR, pL, rhoR, rhoL, eR, eL, cR, cL }
    }
}

File: src/state.jl
module StateContainers
	export State, create_state, InitialConditions, check_state

	struct InitialConditions{T}
		uR::T
		uL::T
		#
		pR::T
		pL::T
		#
		ρR::T
		ρL::T
		#
		eR::T
		eL::T
		#
		cR::T
		cL::T
	end

	function create_ic(gamma::T)::InitialConditions{T} where T <: AbstractFloat
		# these are given in the problem statelemt
		# from the paper
		uL = 0.;
		uR = 0.;
		pL = 10.0^5;
		pR = 10.0^3;
		ρL  = 1.0;
		ρR = 0.01;
		# form the paper p = (gamma - 1) rho e
		# so you can calculate these
		eL = pL / ((gamma - 1) * ρL);
		eR = pR / ((gamma - 1) * ρR);

		# given in the paper again
		cL = sqrt(gamma * pL/ ρL)
		cR = sqrt(gamma * pR / ρR)

		return InitialConditions(uR, uL, pR, pL, ρR, ρL, eR, eL, cR, cL)
	end

	mutable struct State{T}
		gamma::T
		alpha::T
		h::T
		ic::InitialConditions
		u::Vector{T}
		ρ::Vector{T}
		e::Vector{T}
		c::Vector{T}
		E::Vector{T}
		p::Vector{T}
	end

	function create_state(n::Int, h::T, alpha::T)::State{T} where T <: AbstractFloat
		gamma = 1.4

		ic = create_ic(gamma)

		ρ = zeros(n)
		p = zeros(n)
		u = zeros(n)
		e = zeros(n)
		E = zeros(n) # e + 1/2 u^2
		c = zeros(n)

		# the discontinuity should be exactly in the middle
		half_n = Int(floor((n+1) / 2));

		ρ[1:half_n] .= ic.ρL;
		ρ[half_n+1:n] .= ic.ρR;

		p[1:half_n] .= ic.pL;
		p[half_n+1:n] .= ic.pR;

		u[1:half_n] .= ic.uL;
		u[half_n+1:n] .= ic.uR;

		e[1:half_n] .= ic.eL;
		e[half_n+1:n] .= ic.eR;

		E[1:half_n] .= ic.eL + (1/2) .* ic.uL.^2;
		E[half_n+1:n] .= ic.eR + (1/2) .* ic.uR.^2;

		c[1:half_n] .= ic.cL
		c[half_n+1:n] .= ic.cR

		return State(
			gamma,
			alpha,
			h,
			ic,
			u,
			ρ,
			e,
			c,
			E,
			p
		)
	end

	function check_state(s::State)
		if any(isnan, s.ρ)
			error("Nan exists in rho")
		end
		if any(iszero, s.ρ)
			error("zero exists in rho")
		end
		if length(filter(x -> x < 0, s.ρ)) > 0
			display(transpose(s.ρ))
			error("less than zero exists in rho")
		end
		if length(filter(x -> x < 0, s.E)) > 0
			display(transpose(s.E))
			error("less than zero exists in E")
		end
		if length(filter(x -> x < 0, s.e)) > 0
			display(transpose(s.p))
			error("less than zero exists in e")
		end
		if length(filter(x -> x < 0, s.p)) > 0
			display(transpose(s.p))
			error("less than zero exists in p")
		end
		if any(isnan, s.u)
			error("Nan exists in u")
		end
		if any(isnan, s.E)
			error("Nan exists in E")
		end
		if any(isnan, s.e)
			error("Nan exists in e")
		end
		if any(isnan, s.p)
			error("Nan exists in p")
		end
	end
end
